File: examples/languages/nodejs/doom.js
// =============================================================================
//  examples/languages/nodejs/doom.js
// -----------------------------------------------------------------------------
//  Drive the doomkit shared library from Node.js using koffi, the modern Node
//  FFI (think "ctypes for Node"). No native compilation needed. GPLv2. See LICENSE.
// =============================================================================
//
//  The one rule (same as every managed language): keep the registered callbacks
//  reachable so the GC can't collect the function pointers the engine holds. We
//  keep them in the module-level `callbacks` object for the program's lifetime.
//
//  RUN (after `npm install` and building libdoomgeneric -- see bindings/README.md):
//      # put the library where this script looks (./lib) or set DOOMKIT_LIB:
//      cp /path/to/libdoomgeneric.* lib/
//      node doom.js -iwad /path/to/doom1.wad        # writes frame.ppm at frame 100
// =============================================================================

'use strict';

const koffi = require('koffi');
const fs = require('fs');
const path = require('path');

// ---- locate and load the shared library -----------------------------------
const candidates = [
  process.env.DOOMKIT_LIB,
  path.join(__dirname, 'lib', 'libdoomgeneric.dylib'),
  path.join(__dirname, 'lib', 'libdoomgeneric.so'),
  'libdoomgeneric.dylib',
  'libdoomgeneric.so',
].filter(Boolean);

let lib;
for (const c of candidates) {
  try { lib = koffi.load(c); break; } catch (_) { /* try next */ }
}
if (!lib) {
  console.error('could not load libdoomgeneric; build it (make lib) and put it in ./lib or set DOOMKIT_LIB');
  process.exit(1);
}

// ---- describe the C types --------------------------------------------------
// The callback struct: six function pointers (opaque void* to koffi).
const DgCallbacks = koffi.struct('dg_callbacks', {
  init:             'void *',
  draw_frame:       'void *',
  sleep_ms:         'void *',
  get_ticks_ms:     'void *',
  get_key:          'void *',
  set_window_title: 'void *',
});

// Callback prototypes (the signatures C will call back with).
const InitProto   = koffi.proto('void   InitProto()');
const DrawProto   = koffi.proto('void   DrawProto()');
const SleepProto  = koffi.proto('void   SleepProto(uint32_t ms)');
const TicksProto  = koffi.proto('uint32_t TicksProto()');
const GetKeyProto = koffi.proto('int    GetKeyProto(int *pressed, unsigned char *key)');
const TitleProto  = koffi.proto('void   TitleProto(const char *title)');

// The exported library functions.
const dg_set_callbacks = lib.func('void dg_set_callbacks(dg_callbacks *cb)');
const dg_create        = lib.func('void dg_create(int argc, const char **argv)');
const dg_tick          = lib.func('void dg_tick()');
const dg_screen_buffer = lib.func('uint32_t *dg_screen_buffer()');
const dg_resx          = lib.func('int dg_resx()');
const dg_resy          = lib.func('int dg_resy()');

// ---- the six callbacks -----------------------------------------------------
const start = Date.now();
let frame = 0;

function onInit() {
  console.log(`[node] init: framebuffer ${dg_resx()}x${dg_resy()}`);
}

function onDrawFrame() {
  if (++frame !== 100) return;
  const w = dg_resx(), h = dg_resy();
  const ptr = dg_screen_buffer();
  // Copy the framebuffer out of native memory as an array of 0x00RRGGBB words.
  const pixels = koffi.decode(ptr, koffi.array('uint32_t', w * h));

  const rgb = Buffer.allocUnsafe(w * h * 3);
  for (let i = 0; i < w * h; i++) {
    const p = pixels[i];
    rgb[i * 3]     = (p >>> 16) & 0xff;
    rgb[i * 3 + 1] = (p >>> 8) & 0xff;
    rgb[i * 3 + 2] = p & 0xff;
  }
  const out = fs.createWriteStream('frame.ppm');
  out.write(`P6\n${w} ${h}\n255\n`);
  out.end(rgb);
  console.log(`[node] wrote frame.ppm at frame ${frame}`);
}

function onSleepMs(ms) {
  // A tiny busy-wait keeps this single-file demo synchronous. A real app would
  // run the tick loop on a timer instead of blocking.
  const until = Date.now() + ms;
  while (Date.now() < until) { /* spin */ }
}

function onGetTicksMs() { return (Date.now() - start) & 0xffffffff; }
function onGetKey(_pressed, _key) { return 0; }            // no input
function onSetTitle(title) { console.log(`[node] title: ${title}`); }

// Register the JS functions as C callbacks. koffi.register returns a pointer;
// we stash them so they stay alive for the whole run.
const callbacks = {
  init:             koffi.register(onInit,      koffi.pointer(InitProto)),
  draw_frame:       koffi.register(onDrawFrame, koffi.pointer(DrawProto)),
  sleep_ms:         koffi.register(onSleepMs,   koffi.pointer(SleepProto)),
  get_ticks_ms:     koffi.register(onGetTicksMs,koffi.pointer(TicksProto)),
  get_key:          koffi.register(onGetKey,    koffi.pointer(GetKeyProto)),
  set_window_title: koffi.register(onSetTitle,  koffi.pointer(TitleProto)),
};

// ---- run -------------------------------------------------------------------
dg_set_callbacks(callbacks);

const argv = ['doomdemo', ...process.argv.slice(2)]; // prepend a fake argv[0]
dg_create(argv.length, argv);

for (let i = 0; i < 200; i++) dg_tick();
console.log('[node] done.');

File: examples/languages/java/Doom.java
// =============================================================================
//  examples/languages/java/Doom.java
// -----------------------------------------------------------------------------
//  Drive the doomkit shared library from PURE Java using the Foreign
//  Function & Memory API (Project Panama). No JNI, no C glue. Java 22+.
//  GPLv2. See LICENSE.
// =============================================================================
//
//  THE FFM CONCEPTS USED
//    * Linker + SymbolLookup        -> find and call native functions (downcalls)
//    * upcallStub                   -> turn a Java method into a C function ptr
//    * Arena / MemorySegment        -> native memory (the callback struct, argv,
//                                      and reading the framebuffer)
//
//  BUILD & RUN (after building libdoomgeneric -- see bindings/README.md):
//      java --enable-native-access=ALL-UNNAMED \
//           -Djava.library.path=/path/to/lib \
//           Doom.java -iwad doom1.wad
//  (Single-file source launch; needs JDK 22 or newer.)
// =============================================================================

import java.io.FileOutputStream;
import java.lang.foreign.*;
import java.lang.invoke.MethodHandle;
import java.lang.invoke.MethodHandles;
import java.lang.invoke.MethodType;

import static java.lang.foreign.ValueLayout.*;

public final class Doom {

    // One arena for the whole run; it owns the upcall stubs and native buffers,
    // so it must stay open as long as the engine might call back (i.e. forever).
    static final Arena ARENA = Arena.ofShared();
    static final Linker LINKER = Linker.nativeLinker();
    static final SymbolLookup LIB = SymbolLookup.libraryLookup(
            System.mapLibraryName("doomgeneric"), ARENA);

    // ---- downcall handles for the library functions we call ----------------
    static final MethodHandle DG_SET_CALLBACKS = down("dg_set_callbacks", FunctionDescriptor.ofVoid(ADDRESS));
    static final MethodHandle DG_CREATE        = down("dg_create",        FunctionDescriptor.ofVoid(JAVA_INT, ADDRESS));
    static final MethodHandle DG_TICK          = down("dg_tick",          FunctionDescriptor.ofVoid());
    static final MethodHandle DG_SCREEN_BUFFER = down("dg_screen_buffer", FunctionDescriptor.of(ADDRESS));
    static final MethodHandle DG_RESX          = down("dg_resx",          FunctionDescriptor.of(JAVA_INT));
    static final MethodHandle DG_RESY          = down("dg_resy",          FunctionDescriptor.of(JAVA_INT));

    static MethodHandle down(String name, FunctionDescriptor fd) {
        return LINKER.downcallHandle(LIB.find(name).orElseThrow(), fd);
    }

    static final long startMs = System.currentTimeMillis();
    static int frame = 0;

    // ---- the six callbacks, as plain static methods ------------------------

    static void onInit() {
        System.out.printf("[java] init: framebuffer %dx%d%n", resx(), resy());
    }

    static void onDrawFrame() {
        if (++frame != 100) return;
        try {
            int w = resx(), h = resy();
            // dg_screen_buffer returns a zero-length segment; give it a size.
            MemorySegment fb = ((MemorySegment) DG_SCREEN_BUFFER.invoke())
                    .reinterpret((long) w * h * 4);
            byte[] rgb = new byte[w * h * 3];
            for (int i = 0; i < w * h; i++) {
                int p = fb.getAtIndex(JAVA_INT, i);   // 0x00RRGGBB
                rgb[i * 3]     = (byte) (p >> 16);
                rgb[i * 3 + 1] = (byte) (p >> 8);
                rgb[i * 3 + 2] = (byte) p;
            }
            try (var out = new FileOutputStream("frame.ppm")) {
                out.write(("P6\n" + w + " " + h + "\n255\n").getBytes("US-ASCII"));
                out.write(rgb);
            }
            System.out.printf("[java] wrote frame.ppm at frame %d%n", frame);
        } catch (Throwable t) { throw new RuntimeException(t); }
    }

    static void onSleepMs(int ms) {
        try { Thread.sleep(ms & 0xFFFFFFFFL); } catch (InterruptedException ignored) {}
    }
    static int onGetTicksMs() { return (int) (System.currentTimeMillis() - startMs); }
    static int onGetKey(MemorySegment pressed, MemorySegment key) { return 0; } // no input
    static void onSetTitle(MemorySegment title) {
        System.out.println("[java] title: " + title.reinterpret(Long.MAX_VALUE).getString(0));
    }

    // Build a C function pointer from a static method of this class.
    static MemorySegment upcall(String method, FunctionDescriptor fd, MethodType mt) {
        try {
            MethodHandle mh = MethodHandles.lookup().findStatic(Doom.class, method, mt);
            return LINKER.upcallStub(mh, fd, ARENA);
        } catch (ReflectiveOperationException e) { throw new RuntimeException(e); }
    }

    static int resx() { try { return (int) DG_RESX.invoke(); } catch (Throwable t) { throw new RuntimeException(t); } }
    static int resy() { try { return (int) DG_RESY.invoke(); } catch (Throwable t) { throw new RuntimeException(t); } }

    public static void main(String[] args) throws Throwable {
        // Assemble the dg_callbacks struct: 6 consecutive C function pointers.
        MemorySegment cb = ARENA.allocate(ADDRESS.byteSize() * 6);
        cb.setAtIndex(ADDRESS, 0, upcall("onInit",       FunctionDescriptor.ofVoid(),                    MethodType.methodType(void.class)));
        cb.setAtIndex(ADDRESS, 1, upcall("onDrawFrame",  FunctionDescriptor.ofVoid(),                    MethodType.methodType(void.class)));
        cb.setAtIndex(ADDRESS, 2, upcall("onSleepMs",    FunctionDescriptor.ofVoid(JAVA_INT),            MethodType.methodType(void.class, int.class)));
        cb.setAtIndex(ADDRESS, 3, upcall("onGetTicksMs", FunctionDescriptor.of(JAVA_INT),                MethodType.methodType(int.class)));
        cb.setAtIndex(ADDRESS, 4, upcall("onGetKey",     FunctionDescriptor.of(JAVA_INT, ADDRESS, ADDRESS), MethodType.methodType(int.class, MemorySegment.class, MemorySegment.class)));
        cb.setAtIndex(ADDRESS, 5, upcall("onSetTitle",   FunctionDescriptor.ofVoid(ADDRESS),             MethodType.methodType(void.class, MemorySegment.class)));
        DG_SET_CALLBACKS.invoke(cb);

        // Build argv: program name + the user's args, as a C char**.
        MemorySegment argv = ARENA.allocate(ADDRESS.byteSize() * (args.length + 1));
        argv.setAtIndex(ADDRESS, 0, ARENA.allocateFrom("doomdemo"));
        for (int i = 0; i < args.length; i++) {
            argv.setAtIndex(ADDRESS, i + 1, ARENA.allocateFrom(args[i]));
        }
        DG_CREATE.invoke(args.length + 1, argv);

        for (int i = 0; i < 200; i++) DG_TICK.invoke();
        System.out.println("[java] done.");
    }
}
